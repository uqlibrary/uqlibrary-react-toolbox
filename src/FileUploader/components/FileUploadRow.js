import React, {Component} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import FontIcon from 'material-ui/FontIcon';
import IconButton from 'material-ui/IconButton';
import CircularProgress from 'material-ui/CircularProgress';
import {ConfirmDialogBox} from 'uqlibrary-react-toolbox';

class FileUploadRow extends Component {

    static propTypes = {
        index: PropTypes.number.isRequired,
        uploadedFile: PropTypes.object.isRequired,
        onDelete: PropTypes.func,
        locale: PropTypes.object,
        progress: PropTypes.number
    };

    static defaultProps = {
        locale: {
            deleteHint: 'Remove this file',
            deleteRecordConfirmation: {
                confirmationTitle: 'Delete file',
                confirmationMessage: 'Are you sure you want to remove this file from the uploaded queue?',
                cancelButtonLabel: 'No',
                confirmButtonLabel: 'Yes'
            }
        }
    };

    constructor(props) {
        super(props);
    }

    _showConfirmation = () => {
        this.confirmationBox.showConfirmation();
    };

    _deleteFile = () => {
        if (this.props.onDelete) this.props.onDelete(this.props.uploadedFile, this.props.index);
    };

    render() {
        const { deleteRecordConfirmation } = this.props.locale;
        return (
            <div className="columns is-gapless is-mobile uploadedFileRow datalist datalist-row">
                <ConfirmDialogBox
                    onRef={ ref => (this.confirmationBox = ref) }
                    onAction={ this._deleteFile }
                    locale={ deleteRecordConfirmation } />
                <div className="column datalist-text filename">
                    <span className="filename-label">{ this.props.uploadedFile.name }</span>
                </div>
                {
                    this.props.progress === 0 &&
                        <div className="column is-narrow uploadedFileDelete datalist-buttons">
                            <IconButton tooltip={ this.props.locale.deleteHint } onTouchTap={ this._showConfirmation }>
                                <FontIcon className="material-icons deleteIcon">delete</FontIcon>
                            </IconButton>
                        </div>
                }
                {
                    this.props.progress > 0 &&
                        <div className="upload-progress-wrapper">
                            <CircularProgress
                                className="upload-progress"
                                mode="determinate"
                                value={ this.props.progress }
                                size={ 20 }
                                thickness={ 4 }
                            />
                        </div>
                }
                {
                    this.props.progress === 100 &&
                        <FontIcon className="material-icons green-tick">done</FontIcon>
                }
            </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        progress: state.get('fileUpload')[ownProps.uploadedFile.name] || 0
    };
};

export default connect(mapStateToProps)(FileUploadRow);
