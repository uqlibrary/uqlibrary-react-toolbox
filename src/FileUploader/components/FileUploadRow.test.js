import {FileUploadRow} from './FileUploadRow';

function setup(testProps, isShallow = true) {
    const props = {
        ...testProps,
        index: testProps.index || 0,
        uploadedFile: testProps.uploadedFile || {name: 'a.txt', size: 100},
        requireOpenAccessStatus: testProps.requireOpenAccessStatus || false,
        onDelete: testProps.onDelete || jest.fn(),
        onAccessConditionChange: testProps.onAccessConditionChange || jest.fn(),
        onEmbargoDateChange: testProps.onEmbargoDateChange || jest.fn(),
        progress: testProps.progress || 0,
        isUploadInProgress: testProps.isUploadInProgress || false
    };

    return getElement(FileUploadRow, props, isShallow);
}

describe('FileUploadRow', () => {
    it('renders with uploaded file', () => {
        const wrapper = setup({});
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('renders with uploaded file with some progress', () => {
        const wrapper = setup({progress: 50, isUploadInProgress: true});
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('renders for edge browser if file is being uploaded but no progress data', () => {
        const wrapper = setup({progress: 0, isUploadInProgress: true});
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('renders with file upload success', () => {
        const wrapper = setup({progress: 100, isUploadInProgress: true});
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

    it('renders if file uploaded successfully but later other file failed', () => {
        const wrapper = setup({progress: 100, isUploadInProgress: false});
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });
    
    it('call prop to update file metadata with closed access', () => {
        const testFunction = jest.fn();
        const file = new File([""], 'a.txt');
        file.date = '2017-01-01';
        const wrapper = setup({requireOpenAccessStatus: true, onAccessConditionChange: testFunction, uploadedFile: file, index: 0});

        wrapper.instance()._updateAccessCondition(8);
        expect(testFunction).toHaveBeenCalledWith(file, 0, 8);
    });

    it('call prop to update file metadata with open access', () => {
        const testFunction = jest.fn();
        const file = new File([""], 'a.txt');
        file.date = '2017-01-01';
        const wrapper = setup({requireOpenAccessStatus: true, onAccessConditionChange: testFunction, uploadedFile: file, index: 0});

        wrapper.instance()._updateAccessCondition(9);
        expect(testFunction).toHaveBeenCalledWith(file, 0, 9);
    });

    it('call prop to update file metadata with open access date', () => {
        const testFunction = jest.fn();
        const file = new File([""], 'a.txt');
        file.date = '2017-01-01';
        const wrapper = setup({onEmbargoDateChange: testFunction, uploadedFile: file, index: 0});

        wrapper.instance()._updateEmbargoDate(new Date(2016));
        expect(testFunction).toHaveBeenCalledWith(file, 0, new Date(2016));
    });
});
