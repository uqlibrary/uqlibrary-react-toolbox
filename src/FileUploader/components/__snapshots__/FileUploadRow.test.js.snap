// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploadRow renders for edge browser if file is being uploaded but no progress data 1`] = `
<div
  className="columns is-gapless is-multiline uploadedFileRow datalist datalist-row is-clearfix"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to remove this file from the uploaded queue?",
        "confirmationTitle": "Delete file",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text file-info is-6-desktop is-5-tablet is-12-mobile"
  >
    <FontIcon
      className="material-icons mobile-icon is-hidden-desktop is-hidden-tablet"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      attachment
    </FontIcon>
    <div
      className="file-name"
    >
      <span
        className="truncated"
      >
        a.txt
         (
        100.0B
        )
      </span>
      <span
        className="is-mobile label is-hidden-desktop is-hidden-tablet datalist-text-subtitle"
      >
        File name
      </span>
    </div>
  </div>
  <div
    className="column datalist-text is-3-desktop is-4-tablet is-12-mobile"
  />
  <div
    className="column datalist-text is-2-desktop is-2-tablet is-three-quarters-mobile is-inline-block-mobile"
  >
    <div
      className="embargo-date-info"
    />
  </div>
  <div
    className="column is-1-desktop is-1-tablet is-one-quarter-mobile is-inline-block-mobile is-centered is-vcentered"
  >
    <div
      className="upload-progress-info"
    >
      <div
        className="upload-progress"
      >
        <CircularProgress
          max={100}
          min={0}
          mode="indeterminate"
          size={20}
          thickness={4}
          value={0}
        />
      </div>
      <div
        className="upload-progress-number"
      >
        <span
          aria-label="Uploading..."
        >
          Uploading...
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileUploadRow renders if file uploaded successfully but later other file failed 1`] = `
<div
  className="columns is-gapless is-multiline uploadedFileRow datalist datalist-row is-clearfix"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to remove this file from the uploaded queue?",
        "confirmationTitle": "Delete file",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text file-info is-6-desktop is-5-tablet is-12-mobile"
  >
    <FontIcon
      className="material-icons mobile-icon is-hidden-desktop is-hidden-tablet"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      attachment
    </FontIcon>
    <div
      className="file-name"
    >
      <span
        className="truncated"
      >
        a.txt
         (
        100.0B
        )
      </span>
      <span
        className="is-mobile label is-hidden-desktop is-hidden-tablet datalist-text-subtitle"
      >
        File name
      </span>
    </div>
  </div>
  <div
    className="column datalist-text is-3-desktop is-4-tablet is-12-mobile"
  />
  <div
    className="column datalist-text is-2-desktop is-2-tablet is-three-quarters-mobile is-inline-block-mobile"
  >
    <div
      className="embargo-date-info"
    />
  </div>
  <div
    className="column is-1-desktop is-1-tablet is-one-quarter-mobile is-inline-block-mobile is-centered is-vcentered"
  >
    <div
      className="datalist-buttons"
    >
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onTouchTap={[Function]}
        tooltip="Remove this file"
        tooltipPosition="bottom-center"
        touch={false}
      >
        <FontIcon
          className="material-icons deleteIcon"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          delete
        </FontIcon>
      </IconButton>
    </div>
  </div>
</div>
`;

exports[`FileUploadRow renders with file upload success 1`] = `
<div
  className="columns is-gapless is-multiline uploadedFileRow datalist datalist-row is-clearfix"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to remove this file from the uploaded queue?",
        "confirmationTitle": "Delete file",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text file-info is-6-desktop is-5-tablet is-12-mobile"
  >
    <FontIcon
      className="material-icons mobile-icon is-hidden-desktop is-hidden-tablet"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      attachment
    </FontIcon>
    <div
      className="file-name"
    >
      <span
        className="truncated"
      >
        a.txt
         (
        100.0B
        )
      </span>
      <span
        className="is-mobile label is-hidden-desktop is-hidden-tablet datalist-text-subtitle"
      >
        File name
      </span>
    </div>
  </div>
  <div
    className="column datalist-text is-3-desktop is-4-tablet is-12-mobile"
  />
  <div
    className="column datalist-text is-2-desktop is-2-tablet is-three-quarters-mobile is-inline-block-mobile"
  >
    <div
      className="embargo-date-info"
    />
  </div>
  <div
    className="column is-1-desktop is-1-tablet is-one-quarter-mobile is-inline-block-mobile is-centered is-vcentered"
  >
    <div
      className="upload-progress"
    >
      <FontIcon
        className="material-icons green-tick"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        done
      </FontIcon>
    </div>
  </div>
</div>
`;

exports[`FileUploadRow renders with uploaded file 1`] = `
<div
  className="columns is-gapless is-multiline uploadedFileRow datalist datalist-row is-clearfix"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to remove this file from the uploaded queue?",
        "confirmationTitle": "Delete file",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text file-info is-6-desktop is-5-tablet is-12-mobile"
  >
    <FontIcon
      className="material-icons mobile-icon is-hidden-desktop is-hidden-tablet"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      attachment
    </FontIcon>
    <div
      className="file-name"
    >
      <span
        className="truncated"
      >
        a.txt
         (
        100.0B
        )
      </span>
      <span
        className="is-mobile label is-hidden-desktop is-hidden-tablet datalist-text-subtitle"
      >
        File name
      </span>
    </div>
  </div>
  <div
    className="column datalist-text is-3-desktop is-4-tablet is-12-mobile"
  />
  <div
    className="column datalist-text is-2-desktop is-2-tablet is-three-quarters-mobile is-inline-block-mobile"
  >
    <div
      className="embargo-date-info"
    />
  </div>
  <div
    className="column is-1-desktop is-1-tablet is-one-quarter-mobile is-inline-block-mobile is-centered is-vcentered"
  >
    <div
      className="datalist-buttons"
    >
      <IconButton
        disableTouchRipple={false}
        disabled={false}
        iconStyle={Object {}}
        onTouchTap={[Function]}
        tooltip="Remove this file"
        tooltipPosition="bottom-center"
        touch={false}
      >
        <FontIcon
          className="material-icons deleteIcon"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          delete
        </FontIcon>
      </IconButton>
    </div>
  </div>
</div>
`;

exports[`FileUploadRow renders with uploaded file with some progress 1`] = `
<div
  className="columns is-gapless is-multiline uploadedFileRow datalist datalist-row is-clearfix"
>
  <ConfirmDialogBox
    hideCancelButton={false}
    locale={
      Object {
        "cancelButtonLabel": "No",
        "confirmButtonLabel": "Yes",
        "confirmationMessage": "Are you sure you want to remove this file from the uploaded queue?",
        "confirmationTitle": "Delete file",
      }
    }
    onAction={[Function]}
    onRef={[Function]}
  />
  <div
    className="column datalist-text file-info is-6-desktop is-5-tablet is-12-mobile"
  >
    <FontIcon
      className="material-icons mobile-icon is-hidden-desktop is-hidden-tablet"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      attachment
    </FontIcon>
    <div
      className="file-name"
    >
      <span
        className="truncated"
      >
        a.txt
         (
        100.0B
        )
      </span>
      <span
        className="is-mobile label is-hidden-desktop is-hidden-tablet datalist-text-subtitle"
      >
        File name
      </span>
    </div>
  </div>
  <div
    className="column datalist-text is-3-desktop is-4-tablet is-12-mobile"
  />
  <div
    className="column datalist-text is-2-desktop is-2-tablet is-three-quarters-mobile is-inline-block-mobile"
  >
    <div
      className="embargo-date-info"
    />
  </div>
  <div
    className="column is-1-desktop is-1-tablet is-one-quarter-mobile is-inline-block-mobile is-centered is-vcentered"
  >
    <div
      className="upload-progress-info"
    >
      <div
        className="upload-progress"
      >
        <CircularProgress
          max={100}
          min={0}
          mode="determinate"
          size={20}
          thickness={4}
          value={50}
        />
      </div>
      <div
        className="upload-progress-number"
      >
        <span
          aria-label="50%"
        >
          50%
        </span>
      </div>
    </div>
  </div>
</div>
`;
